### PR TITLE
fix structured_output/guided_output for delayed_sampling

### PR DESCRIPTION
**Pull request by inkcherry**
pick https://github.com/HabanaAI/vllm-fork/pull/1270

Under structured_output, Xgrammer needs to modify the logits, and there is a data dependency on the sampling results.
1 if compute logits depend the sampling result , we fetch sample result in advance.
2 if not, we keep its previous behavior to keep performance.
@czhu15

### CODE REVIEW

**Comment by inkcherry** [comment]
> PR looks good, please add a description to explain the existing issue with 'deplayed_sampling + structured_output' and how this PR fixed this issue.

Thanks for pointing that out — I've added a more detailed explanation.
